include/ginkgo/core/solver/solver_base.hpp

```diff
@@ -155,18 +157,95 @@ class ApplyWithInitialGuess {
      */
     void set_default_initial_guess(initial_guess_mode guess) { guess_ = guess; }
 
-    // override this at the same time when overridden
-    virtual void apply_impl(const LinOp* b, LinOp* x) const = 0;
-
-    // override this at the same time when overridden
-    virtual void apply_impl(const LinOp* alpha, const LinOp* b,
-                            const LinOp* beta, LinOp* x) const = 0;
-
 private:
     initial_guess_mode guess_;
 };
 
 
+/**
+ * EnableApplyWithInitialGuess providing default operation for
+ * ApplyWithInitialGuess with correct validation and log. It ensures that
+ * vectors of apply_with_initial_guess will always have the same executor as the
+ * object this mixin is used in, creating a clone on the correct executor if
+ * necessary.
+ *
+ * @tparam DerivedType  The type that this Mixin is used in. It must provide
+ *                      get_size() and get_executor() functions that return
+ *                      correctly initialized values and the logger
+ *                      functionality.
+ */
+template <typename DerivedType>
+class EnableApplyWithInitialGuess : public ApplyWithInitialGuess {
```

Comment:
I add it as TODO. it can do it now, but I don't like the way adding eight if-condition for dense and distributed_vector. Thus, I just put it as virtual now.